src/libteam/patch/0016-block-retry-count-changes.patch

```diff
@@ -0,0 +1,141 @@
+packet comes in without a custom retry count, don't reset it back to
+that custom retry count for 60 seconds since the last new packet.
+---
+ teamd/teamd_runner_lacp.c |   60 +++++++++++++++++++++++++++++++++++++++------
+ 1 file changed, 52 insertions(+), 8 deletions(-)
+
+diff --git a/teamd/teamd_runner_lacp.c b/teamd/teamd_runner_lacp.c
+index 3e8a0f6..c5dad35 100644
+--- a/teamd/teamd_runner_lacp.c
++++ b/teamd/teamd_runner_lacp.c
+@@ -180,6 +180,7 @@ struct lacp {
+ 	struct lacp_port *selected_agg_lead; /* leading port of selected aggregator */
+ 	bool carrier_up;
+ 	time_t warm_start_mode_timer;
++	time_t next_retry_count_change_time;
+ 	struct {
+ 		bool active;
+ #define		LACP_CFG_DFLT_ACTIVE true
+@@ -234,6 +235,7 @@ struct lacp_port {
+ 	struct lacp_port *agg_lead; /* leading port of aggregator.
+ 				     * NULL in case this port is not selected */
+ 	enum lacp_port_state state;
++	int last_received_lacpdu_version;
+ 	bool lacpdu_saved;
+ 	struct lacpdu last_pdu;
+ 	struct {
+@@ -1351,6 +1353,7 @@ static int lacp_port_set_state(struct lacp_port *lacp_port,
+ 		if (err)
+ 			return err;
+ 		lacp_port->lacp->cfg.retry_count = LACP_CFG_DFLT_RETRY_COUNT;
++		lacp_port->last_received_lacpdu_version = 0x01;
+ 		teamd_loop_callback_disable(lacp_port->ctx,
+ 					    LACP_RETRY_COUNT_TIMEOUT_CB_NAME, lacp_port->lacp);
+ 		lacp_port_timeout_set(lacp_port, true);
+@@ -1462,7 +1465,7 @@ static int lacpdu_send(struct lacp_port *lacp_port)
+ 	if (lacp_port->lacp->cfg.enable_retry_count) {
+ 		if (lacp_port->lacp->cfg.retry_count != LACP_CFG_DFLT_RETRY_COUNT
+ 				|| lacp_port->partner_retry_count != LACP_CFG_DFLT_RETRY_COUNT) {
+-			lacpdu_init(&lacpdu, 0xf1);
++			lacpdu_init(&lacpdu, lacp_port->last_received_lacpdu_version);
+ 		} else {
+ 			lacpdu_init(&lacpdu, 0x01);
+ 		}
+@@ -1483,9 +1486,12 @@ static int lacpdu_send(struct lacp_port *lacp_port)
+ 	return err;
+ }
+ 
++#define LACP_RETRY_COUNT_RESET_TIMEOUT_SECONDS 60
++
+ static int lacpdu_process(struct lacp_port *lacp_port, struct lacpdu* lacpdu)
+ {
+ 	int err;
++	struct timespec monotonic_time = {0};
+ 
+ 	if (!lacpdu_check(lacpdu)) {
+ 		teamd_log_warn("malformed LACP PDU came.");
+@@ -1523,17 +1529,55 @@ static int lacpdu_process(struct lacp_port *lacp_port, struct lacpdu* lacpdu)
+ 					lacp_port->partner_retry_count,
+ 					lacpdu->v2.actor_retry_count);
+ 			lacp_port->partner_retry_count = lacpdu->v2.actor_retry_count;
++			if (clock_gettime(CLOCK_MONOTONIC, &monotonic_time)) {
++				err = errno;
++				teamd_log_err("%s: unable to get current time: %s", lacp_port->tdport->ifname, strerror(err));
++				return -err;
++			}
++		}
++		if (lacp_port->partner_retry_count != LACP_CFG_DFLT_RETRY_COUNT) {
++			// Reset the change time every time a 0xf1 packet comes in
++			if (clock_gettime(CLOCK_MONOTONIC, &monotonic_time)) {
++				err = errno;
++				teamd_log_err("%s: unable to get current time: %s", lacp_port->tdport->ifname, strerror(err));
++				return -err;
```

Comment:
What happens to retry count if getting time fails? Will/should this be fatal for teamd?

Reply:
Hmm, with the current code, this would mean the LACPDU packet wouldn't even get processed, which would mean that the 90-second (or however many seconds based on the retry count) timer won't get refreshed, and the LAG would go down.

The timers that teamd uses also use `CLOCK_MONOTONIC`, so I'm inclined to think that if that call fails 3 times, something is messed up with the system.